### PR TITLE
[FIX] handle multiple file origins in ProtID Runs, transfer IDs with CFeat.Obs.

### PR DESCRIPTION
**Pull request by jpfeuffer**
I did not check the tests yet. But this is how it should look like.

Includes some fixed warnings in IsotopeWaveletTransform

### CODE REVIEW

**Comment by jpfeuffer** [comment]
I am not sure if FFID takes those Observations correctly, as it finds less features than on develop when transfer_ids is active.
At least it seems like the overwriting bugs are gone.

**Comment by jpfeuffer** [comment]
Ok I think I might have looked wrongly in the diff of the logs:
https://gist.github.com/jpfeuffer/4d3c696064595a18e31d2c013bf81396
vs
https://gist.github.com/jpfeuffer/4fe58403e2dae06fed4a9e1612901853

I think the actual numbers are always the same. However, if you run PLFQ2 or PLFQ3 (i.e. with transfer_ids) you get different output. Mainly unassigned peptide IDs are missing.

**Comment by jpfeuffer** [comment]
After further investigation the unassigned peptide IDs that are missing are transferred ones. I wonder if there was an error before in that we reported those "fake" IDs as unassigned instead of actually removing them @timosachsenberg

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 1]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
       const String& id = pep.getIdentifier();
       ID::ProcessingStepRef step_ref = id_to_step.at(id);
-      ID::InputFileRef file_ref;
-      if (!step_ref->input_file_refs.empty())
+      ID::InputFileRef inputfile;
+      if (!pep.getBaseName().empty())
       {
-        // @TODO: what if there's more than one input file?
-        file_ref = step_ref->input_file_refs[0];
+        const auto file_ref = id_data.registerInputFile(ID::InputFile(pep.getBaseName()));
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]: just: inputfile = id_data... ?

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 2]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
+        inputfile = file_ref;
       }
       else
       {
-        String file = "UNKNOWN_INPUT_FILE_" + id;
-        file_ref = id_data.registerInputFile(ID::InputFile(file));
+        if (!step_ref->input_file_refs.empty())
+        {
+          if (step_ref->input_file_refs.size() > 1)
+          {
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]:
```suggestion
          { // Undo the hack needed in the legacy id datastructure to represent merged id files. Extract the actual input file name so we can properly register it.
```

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 3]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
+                  );
+            }
+          }
+          else // one file in the ProteinIdentification Run only
+          {
+            inputfile = step_ref->input_file_refs[0];
+          }
+        }
+        else
+        {
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]:
```suggestion
        { // no input file annotated in legacy data structure
```

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 4]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
+          String file = "UNKNOWN_INPUT_FILE_" + id;
+          ID::InputFileRef file_ref =
+              id_data.registerInputFile(ID::InputFile(file));
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]: just inputfile = id_data.register... ?

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 5]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
+          inputfile = file_ref;
+        }
       }
-      ID::Observation obs("", file_ref); // fill in "data_id" later
-      obs.rt = pep.getRT();
-      obs.mz = pep.getMZ();
-      obs.addMetaValues(pep);
+      String data_id;
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]:
```suggestion
      String data_id; // an identifier unique to the input file
```

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp [hunk 6]
```diff
@@ -201,39 +201,70 @@ namespace OpenMS
...
       if (pep.metaValueExists("spectrum_reference"))
       {
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]:
```suggestion
      {  // use spectrum native id if present
```

File: src/utils/ProteomicsLFQ.cpp [hunk 7]
```diff
@@ -1076,16 +1078,26 @@ class ProteomicsLFQ :
         MapAlignmentTransformer::transformRetentionTimes(peptide_ids, transformations[fraction_group - 1]);
         MapAlignmentTransformer::transformRetentionTimes(ms_centroided, transformations[fraction_group - 1]);
 
+        //consensus_fraction.getIdentifier and getLoadedFilePath seems to be empty
+        auto tmpfileref = id_data.registerInputFile(IdentificationData::InputFile("fraction"+String(fraction)+".consensusXML")); // "temporary" ConsensusXML for this fraction
```
Comment (timosachsenberg) [inline src/utils/ProteomicsLFQ.cpp]:
```suggestion
        auto tmpfileref = id_data.registerInputFile(IdentificationData::InputFile("fraction" + String(fraction) + ".consensusXML")); // "temporary" ConsensusXML for this fraction
```

**Review by timosachsenberg — APPROVED** [review]
Keeping the PSMs of the transfered ones was indeed a bug. Thanks for fixing.

**Comment by jpfeuffer** [comment]
So, after fixing a bug found by Timo where I forgot to register the ObsMatches for the transferred IDs in the end,
I thought we would need to remove those IDs later on to not be exported.
But I found out that removing the transferred IDs is not necessary since they do not have a score and ObsMatches without scores are not exported.
I left it like this for now and added an option to include those matches without score (which is by default deactivated).

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h [hunk 8]
```diff
@@ -1362,6 +1362,15 @@ namespace OpenMS
 
     /// @name Filter functions for class IdentificationData
     ///@{
+
+    template <typename PredicateType>
+    static void filterObservationMatchesByFunctor(
+        IdentificationData& id_data, PredicateType&& func, bool cleanup_affected = false)
+    {
+      id_data.removeFromSetIf_(id_data.observation_matches_, func);
+      if (cleanup_affected) id_data.cleanup();
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:1371]: Is false a good default?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp [hunk 9]
```diff
@@ -282,11 +282,14 @@ namespace OpenMS
...
     {
-      if (ref->getMetaValue("FFId_category", "") == "seed")
+      const String& cat = ref->getMetaValue("FFId_category", "");
+      if (cat == "seed")
       {
         n_seed_targets_++;
       }
-      else
+      else if (cat == "transfer")
+        {std::cout << "TRANSFER" << std::endl;}
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp]: /n

Comment (jpfeuffer) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderIdentificationAlgorithm.cpp]: Oh this should go completely.

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h (hunk 8 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:1371]: I am not sure. id_data.cleanup() has 5 optional bools for removing related stuff.
I am not sure if we want to remove the related Observation and the IdentifiedPeptide if this ObsMatch is removed. We need to think about it.

File: src/openms/source/METADATA/ID/IdentificationDataConverter.cpp (hunk 1 above)
Comment (jpfeuffer) [reply src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]: can you elaborate?

Comment (timosachsenberg) [reply src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]: line https://github.com/hendrikweisser/OpenMS/pull/16/files/4346c892369fdd264045dba41ebb04d0b9f7c0f3#diff-eb57500eeb186958ca2acd00517194d8b96446a4b25002b0403d3c361268b378R208 seems redundant

Comment (jpfeuffer) [reply src/openms/source/METADATA/ID/IdentificationDataConverter.cpp]: oh yes. maybe I used it for some debug output inbetween

**Comment by jpfeuffer** [comment]
This is kind of finished from my side. I need input for the cleanup. And a general review from @hendrikweisser I think.
We then need to open a PR to OpenMS dev to see the diffs for the test files to see if they make sense.
But probably we need quite some testing for real-life data. Since we cannot really test for equality anymore (due to different ordering and now also the removal of those transfer IDs, plus some general non-deterministic behaviour [maybe during fitting?]), we probably need to test qualitatively.

**Comment by hendrikweisser** [comment]
I had to handle this quickly before packing my bags. PR against "develop" is here: https://github.com/OpenMS/OpenMS/pull/5352
Let's discuss everything else there.

**Pull request merged**